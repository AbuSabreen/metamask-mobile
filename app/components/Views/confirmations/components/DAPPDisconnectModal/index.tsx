import Modal from 'react-native-modal';
import React, { useCallback, useState } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';

import { strings } from '../../../../../../locales/i18n';
import Button, {
  ButtonSize,
  ButtonVariants,
} from '../../../../../component-library/components/Buttons/Button';
import ButtonIcon, {
  ButtonIconVariants,
} from '../../../../../component-library/components/Buttons/ButtonIcon';
import Device from '../../../../../util/device';
import Icon from '../../../../../component-library/components/Icons/Icon/Icon';
import {
  IconColor,
  IconName,
  IconSize,
} from '../../../../../component-library/components/Icons/Icon';
import TagUrl from '../../../../../component-library/components/Tags/TagUrl';
import { removePermittedAccounts } from '../../../../../core/Permissions';
import { resetRejectionToRequestFromOrigin } from '../../../../../actions/requests';
import { selectSelectedAddress } from '../../../../../selectors/preferencesController';
import { useFavicon } from '../../../../../components/hooks/useFavicon';
import { useTheme } from '../../../../../util/theme';

import useRejectionToRequestFromOriginInfo from '../../hooks/useRejectionToRequestFromOriginInfo';

const createStyles = (colors: any) =>
  StyleSheet.create({
    successContainer: {
      backgroundColor: colors.background.default,
      paddingTop: 24,
      minHeight: '30%',
      borderTopLeftRadius: 20,
      borderTopRightRadius: 20,
      paddingBottom: Device.isIphoneX() ? 20 : 0,
      alignItems: 'center',
      justifyContent: 'space-between',
    },
    container: {
      backgroundColor: colors.background.default,
      paddingTop: 24,
      minHeight: '60%',
      borderTopLeftRadius: 20,
      borderTopRightRadius: 20,
      paddingBottom: Device.isIphoneX() ? 20 : 0,
      alignItems: 'center',
      justifyContent: 'space-between',
    },
    buttonContainer: {
      display: 'flex',
      flexDirection: 'row',
      paddingVertical: 24,
      paddingHorizontal: 24,
    },
    buttonDivider: {
      width: 8,
    },
    button: {
      flex: 1,
      marginHorizontal: 8,
    },
    bottomModal: {
      justifyContent: 'flex-end',
      margin: 0,
    },
    title: {
      fontWeight: '700',
      fontSize: 16,
      color: colors.text.default,
    },
    successTitle: {
      fontWeight: '700',
      fontSize: 16,
      color: colors.text.default,
      marginVertical: 24,
    },
    info: {
      fontWeight: '400',
      fontSize: 14,
      color: colors.text.default,
      textAlign: 'center',
      width: '90%',
    },
    tagUrl: {
      marginTop: 8,
      marginBottom: 8,
    },
    closeIcon: {
      position: 'absolute',
      right: 20,
      top: 20,
    },
  });

const DAPPDisconnectModal = () => {
  const { blockedOrigins } = useRejectionToRequestFromOriginInfo();
  const selectedAddress = useSelector(selectSelectedAddress);
  const dispatch = useDispatch();
  const { colors } = useTheme();
  const styles = createStyles(colors);
  const faviconSource = useFavicon(blockedOrigins?.[0]);
  const [siteDisconnected, setSiteDisconnected] = useState(false);

  const onCancel = useCallback(
    () => dispatch(resetRejectionToRequestFromOrigin(blockedOrigins[0])),
    [blockedOrigins, dispatch],
  );

  const onDisconnect = useCallback(() => {
    dispatch(resetRejectionToRequestFromOrigin(blockedOrigins[0]));
    removePermittedAccounts(blockedOrigins[0], [selectedAddress]);
    setSiteDisconnected(true);
  }, [blockedOrigins, dispatch, selectedAddress]);

  const onModalClose = useCallback(() => {
    if (siteDisconnected) {
      setSiteDisconnected(false);
    } else {
      onCancel();
    }
  }, [setSiteDisconnected, siteDisconnected, onCancel]);

  if (!blockedOrigins.length && !siteDisconnected) {
    return null;
  }

  return (
    <Modal
      isVisible
      animationIn="slideInUp"
      animationOut="slideOutDown"
      style={styles.bottomModal}
      backdropColor={colors.overlay.default}
      backdropOpacity={1}
      animationInTiming={600}
      animationOutTiming={600}
      onBackButtonPress={onModalClose}
      onSwipeComplete={onModalClose}
      swipeDirection={'down'}
      propagateSwipe
    >
      {siteDisconnected ? (
        <View style={styles.successContainer}>
          <Icon
            name={IconName.CheckBoxOn}
            color={IconColor.Success}
            size={IconSize.Xl}
          />
          <ButtonIcon
            variant={ButtonIconVariants.Secondary}
            style={styles.closeIcon}
            iconName={IconName.Close}
            onPress={() => setSiteDisconnected(false)}
          />
          <Text style={styles.successTitle}>
            {strings('disconnect_modal.success_title')}
          </Text>
          <Text style={styles.info}>
            {strings('disconnect_modal.success_info')}
          </Text>
          <View style={styles.buttonContainer}>
            <Button
              variant={ButtonVariants.Primary}
              onPress={() => setSiteDisconnected(false)}
              label={strings('disconnect_modal.got_it')}
              size={ButtonSize.Lg}
              style={styles.button}
            />
          </View>
        </View>
      ) : (
        <View style={styles.container}>
          <Icon
            name={IconName.Danger}
            color={IconColor.Warning}
            size={IconSize.Xl}
          />
          <Text style={styles.title}>{strings('disconnect_modal.title')}</Text>
          <TagUrl
            imageSource={faviconSource}
            label={blockedOrigins[0]}
            style={styles.tagUrl}
          />
          <Text style={styles.info}>{strings('disconnect_modal.info_1')}</Text>
          <Text style={styles.info}>{strings('disconnect_modal.info_2')}</Text>
          <View style={styles.buttonContainer}>
            <Button
              variant={ButtonVariants.Secondary}
              onPress={onCancel}
              label={strings('disconnect_modal.cancel_cta')}
              size={ButtonSize.Lg}
              style={styles.button}
            />
            <View style={styles.buttonDivider} />
            <Button
              variant={ButtonVariants.Primary}
              onPress={onDisconnect}
              label={strings('disconnect_modal.disconnect_cta')}
              size={ButtonSize.Lg}
              style={styles.button}
            />
          </View>
        </View>
      )}
    </Modal>
  );
};

export default DAPPDisconnectModal;
