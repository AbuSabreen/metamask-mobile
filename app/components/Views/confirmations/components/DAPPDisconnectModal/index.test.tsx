import React from 'react';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import initialBackgroundState from '../../../../../util/test/initial-background-state.json';
import renderWithProvider from '../../../../../util/test/renderWithProvider';

import DAPPDisconnectModal from './index';

const mockStore = configureMockStore();

const initialState = {
  engine: {
    backgroundState: initialBackgroundState,
  },
};

const store = mockStore(initialState);

jest.mock('../../hooks/useRejectionToRequestFromOriginInfo', () => () => ({
  blockedOrigins: ['www.test.com'],
}));

function render() {
  return renderWithProvider(
    <Provider store={store}>
      <DAPPDisconnectModal />
    </Provider>,
  );
}

describe('DAPPDisconnectModal', () => {
  it('should render correctly', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });
});
