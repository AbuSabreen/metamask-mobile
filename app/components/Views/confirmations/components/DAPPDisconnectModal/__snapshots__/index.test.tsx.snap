// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DAPPDisconnectModal should render correctly 1`] = `
<Modal
  animationType="none"
  deviceHeight={null}
  deviceWidth={null}
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onBackdropPress={[Function]}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  onSwipeComplete={[Function]}
  panResponderThreshold={4}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  statusBarTranslucent={false}
  supportedOrientations={
    [
      "portrait",
      "landscape",
    ]
  }
  swipeDirection="down"
  swipeThreshold={100}
  transparent={true}
  visible={true}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    forwardedRef={[Function]}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#00000099",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    collapsable={false}
    deviceHeight={null}
    deviceWidth={null}
    forwardedRef={[Function]}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onSwipeComplete={[Function]}
    panResponderThreshold={4}
    pointerEvents="box-none"
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    style={
      {
        "flex": 1,
        "justifyContent": "flex-end",
        "left": 0,
        "margin": 0,
        "top": 0,
        "transform": [
          {
            "translateY": 1334,
          },
        ],
      }
    }
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection="down"
    swipeThreshold={100}
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderTopLeftRadius": 20,
          "borderTopRightRadius": 20,
          "justifyContent": "space-between",
          "minHeight": "60%",
          "paddingBottom": 20,
          "paddingTop": 24,
        }
      }
    >
      <SvgMock
        color="#BF5208"
        height={32}
        name="Danger"
        style={
          {
            "height": 32,
            "width": 32,
          }
        }
        width={32}
      />
      <Text
        style={
          {
            "color": "#24272A",
            "fontSize": 16,
            "fontWeight": "700",
          }
        }
      >
        Potential scam detected
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#FFFFFF",
            "borderColor": "#BBC0C5",
            "borderRadius": 99,
            "borderWidth": 1,
            "flexDirection": "row",
            "marginBottom": 8,
            "marginTop": 8,
            "paddingLeft": 8,
            "paddingRight": 16,
            "paddingVertical": 8,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 16,
              "height": 24,
              "marginRight": 8,
              "overflow": "hidden",
              "width": 24,
            }
          }
        >
          <Image
            onError={[Function]}
            resizeMode="contain"
            source={{}}
            style={
              {
                "flex": 1,
                "height": undefined,
                "width": undefined,
              }
            }
            testID="favicon-avatar-image"
          />
        </View>
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#535A61",
              "flexShrink": 1,
              "fontFamily": "Euclid Circular B",
              "fontSize": 14,
              "fontWeight": "400",
              "letterSpacing": 0,
              "lineHeight": 22,
            }
          }
        >
          www.test.com
        </Text>
      </View>
      <Text
        style={
          {
            "color": "#24272A",
            "fontSize": 14,
            "fontWeight": "400",
            "textAlign": "center",
            "width": "90%",
          }
        }
      >
        The site has sent you multiple requests. It might be a scam site trying to steal your assets by sending multiple requests.
      </Text>
      <Text
        style={
          {
            "color": "#24272A",
            "fontSize": 14,
            "fontWeight": "400",
            "textAlign": "center",
            "width": "90%",
          }
        }
      >
        We recommend disconnecting from this site.
      </Text>
      <View
        style={
          {
            "display": "flex",
            "flexDirection": "row",
            "paddingHorizontal": 24,
            "paddingVertical": 24,
          }
        }
      >
        <TouchableOpacity
          accessibilityRole="button"
          accessible={true}
          activeOpacity={1}
          onPress={[Function]}
          onPressIn={[Function]}
          onPressOut={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "transparent",
              "borderColor": "#0376C9",
              "borderRadius": 24,
              "borderWidth": 1,
              "flex": 1,
              "flexDirection": "row",
              "height": 48,
              "justifyContent": "center",
              "marginHorizontal": 8,
              "paddingHorizontal": 16,
            }
          }
          variant="Secondary"
        >
          <Text
            accessibilityRole="text"
            style={
              {
                "color": "#0376C9",
                "fontFamily": "Euclid Circular B",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 22,
              }
            }
          >
            Cancel
          </Text>
        </TouchableOpacity>
        <View
          style={
            {
              "width": 8,
            }
          }
        />
        <TouchableOpacity
          accessibilityRole="button"
          accessible={true}
          activeOpacity={1}
          onPress={[Function]}
          onPressIn={[Function]}
          onPressOut={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "#0376C9",
              "borderRadius": 24,
              "flex": 1,
              "flexDirection": "row",
              "height": 48,
              "justifyContent": "center",
              "marginHorizontal": 8,
              "paddingHorizontal": 16,
            }
          }
          variant="Primary"
        >
          <Text
            accessibilityRole="text"
            style={
              {
                "color": "#FFFFFF",
                "fontFamily": "Euclid Circular B",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 22,
              }
            }
          >
            Disconnect
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  </View>
</Modal>
`;
